### DISCUSSION TITLE
clarification on constraints for installation vs upgrade

### DISCUSSION THREAD

**Discussion by cburroughs** [apache/airflow#28732] [Q&A] [answered]
I'm working with a setup that makes significant use of the `PythonOperator` in a managed environment (that is, Airflow itself is already installed).  Additional requirements for our code are deployed with a simple `pip install -r requirements.txt` on top of the existing Airflow install.  After reading about constraints files <https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html> I became concerned that this was incorrect, or at least only accidentally worked.  My interpretation of the docs was that the constraints represented something akin to a java servlet container or system C libraries -- an immutable fact of the environment that ought not to be changed by user code.

In trying to get the `requirements.txt` file to abide by the published Airflow constraints I found -- perhaps not surprisingly! -- many conflicts.  And in trying to resolve the conflicts I cam across discussion of how the constraints files were expected to be used that didn't match my initial reading of the docs:
 * https://github.com/apache/airflow/issues/17584
 * https://github.com/apache/airflow/discussions/19236
  
  In particular: *once airflow is installed, users should be able to upgrade any of the dependencies if new versions are released*
  
 I'm now unsure on the proper use of Airflow constraints files.  Do they:
  * (A) Constrain everything in the Airflow environment.  If the constraint says `pyOpenSSL==20.0.1` then only that version may be used.
   * (B)  Constraint only the initially reproducible *installation*.  After that, additional libraries or upgrades for user code (ex to `pyOpenSSL`) need only respect Airflow's transitive `install_requires`.

**Comment by potiuk** [comment, marked as answer]
It's B)

If you think you can update the docs to be clearer about it - feel free to make PEa to fix the parts the confused you. 

It is super simple - just click 'suggest a change on this page' and you will get PR openes where you will be able to propose changes. 

Happy to review it and merge after you open it.
